### DISCUSSION TITLE
Best way to track memory usage during a simulation?

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#1175] [Q&A] [answered]
There are some different tools for tracking memory usage during a simulation, for example:
- vmstat
- dstat
- free

Which is the best tool to use? Do these tools track memory usage identically? 

@jtracey IIRC, there is some important aspect of the memory counting that `free` does correctly but the other tools do not do correctly. Could you explain the situation w.r.t. tracking memory, and offer guidance on using one or more of these tools. (My goal is to get the details here so we don't have to rely on my memory.) Thank you!

**Comment by sporksmith** [comment]
I also recently learned of `pmap`, for digging further into the usage for a specific process. Notably it includes PSS for each mapping, which is helpful for understanding shared memory regions.

There are also some useful recipes using `perf` and `bpf` to track calls to malloc (user-space allocation) and/or page fault events (when the OS is forced to actually assign a physical page to a virtual page) and the stacks at which they occur. As with pmap these are more useful for digging into details of how memory is getting allocated and used than as a first line tool for measuring usage.

**Comment by jtracey** [comment, marked as answer]
Yeah, the trick with measuring memory usage of a process on Linux is that there is no one best way, since "memory usage of a process" isn't actually a well-defined thing. There is physical memory, and the kernel manages it for processes, often in ways that don't directly correspond to the virtual address space, or even explicit mappings. The two most notable issues are [copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write) and [same-page merging](https://en.wikipedia.org/wiki/Kernel_same-page_merging), with memory compression being another less common one, and kernel memory used while managing a process being a typically negligible one in Shadow's case (but not for, e.g., NetMirage). There are some measurements that try to mitigate this, like `smem` and `pmap`'s pss (proportional set size) that @sporksmith mentioned, which just takes memory from shared pages and divides it by the number of processes sharing it, but that's still not really measuring memory being "used" by a process, since you can have pages shared for unrelated purposes (killing a process using one share doesn't free that share, it will just start being reported as used by other processes; starting a process that uses the same shared libraries as a bunch of other unrelated processes will report usage of that shared library as a small bump to its pss, without increasing memory used on the system, and without representing the cost if those other processes weren't running). Ultimately there's no answer to the question of "if two processes are using transparently shared pages, which of them are *really* using the memory, and how much?"

What we've done in the past instead is to measure the amount of *available* memory on the system, and monitor how that changes. The advantage of that is it tells you what you ultimately care about: how big of an experiment can I run before the kernel kills it? The main disadvantage is it's pretty noisy, from everything else going on in your system. In our case the noise has been negligible enough that we can use it, since Shadow was pretty much always using far, far more memory than anything else on the system. But for smaller experiments, this measurement wouldn't be as accurate (you're also unlikely to trigger same-page merging if you're not squeezing the kernel for memory, so if there's lots of identical state somewhere, it will look like you're using more memory than you actually need). The other disadvantage is you can't really use it to profile, only benchmark, since it obviously won't tell you where the memory is being used. So which measurement is best depends on what purpose you're tracking memory usage for.

**Reply by robgjansen** [reply]
Thank you for the detailed answer!

**Reply by robgjansen** [reply]
Just a note that if `free -w` provides an `available` column on your system, then the best way to calculate RAM used seems to be: `total - available`. Ironically, we found that the `used` column may not account for all RAM actually in use in all cases.
